### PR TITLE
[LiveComponent] fix required select not initialized

### LINKED ISSUE #1332: [UX] Live component - form select required  not initialized
### Symfony version(s) affected

6.3.10

### Description

Class: Symfony\UX\LiveComponent\ComponentWithFormTrait::extractFormValues

Doesn't extract required values for select / entity type field:

image
image

The default value should be '1'.

This is the FormType:
```php
            ->add('project', EntityType::class, [
                'class' => Project::class,
                'label' => 'label.project',
                'required' => true,
                'multiple' => false,
                'choice_label' => fn ($project) => $project->getName(),
            ])
```

### How to reproduce

Create a form type with ux live component and required entity type field. 
https://symfony.com/bundles/ux-live-component/current/index.html#forms

### Possible Solution

If select / entity type and required, extract the default value

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dsoriano**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | Fix #1332 
| License       | MIT

Initialize required select(ChoiceType or EntityType) with the first entry. Not applied to multiple or expanded.

### CODE REVIEW

**Comment by dsoriano** [comment]
If someone knows why ux.symfony.com tests failed... it's gives me headache.. Nothing was changed in the ux.symfony.com directory and in local the tests are ok.

**Comment by smnandre** [comment]
It appears this PR changes are indeed responsible...

Capture d’écran 2024-11-22 à 17 49 54


The ux.symfony.com tests on the CI use the UX packages of the PR... 

Locally to reproduce you should go into the ux.symfony.com repository and then call the link locally command


```
cd ux.symfony.com
php bin/link-locally
```

Then refresh 

```
composer update
php bin/console c:c
symfony server:start -d

```

**Comment by dsoriano** [comment]
ok great, thank you ! I will take a look on it.
Weird that locally the tests for ux.symfony.com are ok, but I probably missed something...

Thanks again

**Comment by smnandre** [comment]
You did not run the php bin/link-locally i suppose ? 

I let you look at the suggestion, i found the source of the problem. 

I let you check locally and tell me if that seems ok for you ?

File: src/LiveComponent/src/ComponentWithFormTrait.php [hunk 1]
```diff
@@ -258,6 +258,20 @@ private function extractFormValues(FormView $formView): array
...
+                && !$child->vars['multiple']
+                && !$child->vars['expanded']
+                && null === ($values[$name] ?? null)
+            ) {
+                if (0 === \count($choices = $child->vars['choices'])) {
+                    throw new UnprocessableEntityHttpException(\sprintf('The required field "%s" is missing and has no default value', $name));
+                }
+
+                $values[$name] = reset($choices)->value;
             } else {
```
Comment (smnandre) [inline src/LiveComponent/src/ComponentWithFormTrait.php]:
```suggestion
                continue;
            }

           // Special handling: required <select> fields with no value, no
            // placeholder and no multiple attribute should have their value set
            // to the first choice, if it exists.
            if (
                \array_key_exists('choices', $child->vars)
                && $child->vars['required']
                && !$child->vars['disabled']
                && !$child->vars['value']
                && !$child->vars['placeholder']
                && !$child->vars['multiple']
                && !$child->vars['expanded']
                && $child->vars['choices']
            ) {
                if (null !== $firstKey = array_key_first($child->vars['choices'])) {
                    $values[$name] = $child->vars['choices'][$firstKey]->value ?? null;
                    
                    continue;
                }
            }
            
```

There were some pre-condition checks missing. This one seems more robust, i'll let you check.

File: src/LiveComponent/tests/Fixtures/Form/FormWithManyDifferentFieldsType.php [hunk 2]
```diff
@@ -41,6 +43,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
                 ],
+                'required' => false,
+            ])
+            ->add('choice_required', ChoiceType::class, [
+                'choices' => [
+                    'foo' => 1,
+                    'bar' => 2,
+                ],
             ])
             ->add('choice_expanded', ChoiceType::class, [
```
Comment (smnandre) [inline src/LiveComponent/tests/Fixtures/Form/FormWithManyDifferentFieldsType.php]:
```suggestion
                'required' => false,
            ])
            ->add('choice_required_with_placeholder', ChoiceType::class, [
                'choices' => [
                    'bar' => 2,
                    'foo' => 1,
                ],
                'required' => true,
                'placeholder' => 'foo',
            ])
            ->add('choice_required_without_placeholder', ChoiceType::class, [
                'choices' => [
                    'bar' => 2,
                    'foo' => 1,
                ],
                'required' => true,
                'placeholder' => false,
            ])
            ->add('choice_expanded', ChoiceType::class, [
```

I added a test to distinct the two scenarios

File: src/LiveComponent/tests/Functional/Form/ComponentWithFormTest.php [hunk 3]
```diff
@@ -174,9 +178,11 @@ public function testHandleCheckboxChanges(): void
             'textarea' => '',
             'range' => '',
             'choice' => '',
+            'choice_required' => '1',
             'choice_expanded' => '',
             'choice_multiple' => ['2'],
             'select_multiple' => ['2'],
+            'entity' => (string) $id,
```
Comment (smnandre) [inline src/LiveComponent/tests/Functional/Form/ComponentWithFormTest.php]:
```suggestion
            'choice_required_with_placeholder' => '',
            'choice_required_without_placeholder' => '2',
            'choice_expanded' => '',
            'choice_multiple' => ['2'],
            'select_multiple' => ['2'],
            'entity' => (string) $id,
```

**Comment by dsoriano** [comment]
@smnandre indeed I missed the link-locally, thanks for your help. Suggestions was good, all is ok now. Many thanks !

**Review by smnandre — APPROVED** [review]
Thank you very much @dsoriano!

**Comment by Kocal** [comment]
Thank you Denis and Simon :)

**Pull request merged**